src/Symfony/Component/Messenger/Envelope.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger;
+
+/**
+ * A message wrapped in an envelope with items (configurations, markers, ...).
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.1
+ */
+final class Envelope
+{
+    private $items = array();
+    private $message;
+
+    /**
+     * @param object         $message
+     * @param EnvelopeItem[] $items
+     */
+    public function __construct($message, array $items = array())
+    {
+        $this->message = $message;
+        foreach ($items as $item) {
+            $this->with($item);
+        }
+    }
+
+    /**
+     * Wrap a message into an envelope if not already wrapped.
+     *
+     * @param Envelope|object $message
+     */
+    public static function wrap($message): self
+    {
+        return $message instanceof self ? $message : new self($message);
+    }
+
+    public function with(EnvelopeItem $item): self
+    {
+        $this->items[\get_class($item)] = $item;
+
+        return $this;
+    }
+
+    public function get(string $itemFqcn): ?EnvelopeItem
+    {
+        return $this->items[$itemFqcn] ?? null;
+    }
+
+    /**
+     * @return EnvelopeItem[] indexed by fqcn
+     */
+    public function all(): array
+    {
+        return $this->items;
+    }
+
+    /**
+     * @return object The original message contained in the envelope
+     */
+    public function open()
```

Comment:
It doesn't feel like a pure getter method... what about `message()` or `getMessage()`?

Reply:
Was the previous naming. No strong opinion.

Reply:
3 against 1. Changed back to `Envelope::getMessage()`.

I liked the "open" semantic but I understand it might seem weird as first sight and might expect side effects.